### PR TITLE
Prepare sandbox:3.0 stable

### LINKED ISSUE #614: Upgrade to Symfony 3.4
Hello,

My name is Wojciech Błoszyk, a PHP developer from Poland. I work with sonata-project bundles for many years.

I want upgrade sandbox for 3.0 (stable) to use Symfony 3.4 (minimum require for ecommerce 3.x)

Work i done for now:
- [x] create new symfony project
- [x] configure require bundles
- [x] extends bundles
- [x] fix bin/load_data.php and fixtures (without LoadOrderData)
- [x] quick fix blocks to run alfa version with bugs

Work to do:
- [x] run Test
- [x] add User Block
- [x] fix TwitterEmbedTweetBlockService
- [x] fix LoadOrderData Fixtures
- [x] upgrade behat to 3.4
- [x] config and check API
- [x] update documentation
- [x] change assetic to assets or symfony/webpack-encore
- [ ] config travis

Also i have questions:
1. Should I renew User Block in SonataUserBundle or create it in SonataDemoBundle?
1. Could someone help me with Test and travis config?
1. Do u have any roadmap for sonata bundles?

[link to my fork - branch 3.0](https://github.com/wbloszyk/sandbox/tree/3.0)

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

This PR contains:
- fixes from php-cs-fix (from dev-kit)
- update Symfony
    - fix structure
    - add autowiring 
- update behat to ^3.4
- change assetic to assets or symfony/webpack-encore
    - jquery (set as global)
    - bootstrap
    - select2
    - other js is set as asset (to change in sandbox:4.x)
- reorganize profile  
- override ApplicationFOSUserBundle templates 
- add type hints
- update README.md
- and many more

Closes #614

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,60 +13,74 @@ Sonata Standard Edition comes pre-configured with the following bundles:
...
+    git checkout 3.0
+    rm -r .git
+    
+Configure default the parameters.yml file: 
+
+    cp app/config/parameters.yml.dist app/config/parameters.yml
+        
+Download vendors:
+
+    php composer.phar update
```
Comment (core23) [inline README.md]:
```suggestion
    composer update
```

File: README.md [hunk 2]
```diff
@@ -13,60 +13,74 @@ Sonata Standard Edition comes pre-configured with the following bundles:
...
+--------------------
+* vagrant up --provision --provider=virtualbox (Vagrant is going to get the environnement, install it for you and load sonata sample data)
+* Configure your host ``sudo nano /etc/hosts`` and add this line ``192.168.33.99   sonata.local``
+* Open your browser [here][link_sonata]
+
 Run
 ---
 
-If you are running PHP5.4, you can use the built in server to start the demo:
+If you are running PHP >= 5.5.9 or >=7.0.8, you can use the built in server to start the demo:
```
Comment (core23) [inline README.md]: Nobody should use PHP5 anymore

File: app/config/sonata/sonata_block.yml [hunk 3]
```diff
@@ -55,13 +55,9 @@ sonata_block:
         sonata.basket.block.nb_items:
```
Comment (core23) [inline app/config/sonata/sonata_block.yml]: You can remove this. The blocks services are auto-registered

File: composer.json [hunk 4]
```diff
@@ -5,92 +5,92 @@
...
-
-        "sensio/distribution-bundle": "^5.0.21",
-        "sensio/framework-extra-bundle": "^5.1.5",
-        "sensio/generator-bundle": "^3.1.7",
-
-        "jms/serializer-bundle": "^1.5.0",
-
-        "doctrine/doctrine-bundle": "^1.8.1",
-        "doctrine/doctrine-migrations-bundle": "^1.3.1",
+        "php": ">=5.5.9",
```
Comment (core23) [inline composer.json]:
```suggestion
        "php": "^7.1",
```

File: README.md (hunk 1 above)
Comment (wbloszyk) [reply README.md]: I update it by undo commits to avoid unnessesery commits.

File: README.md (hunk 2 above)
Comment (wbloszyk) [reply README.md]: Yes, PHP >= 7.1 (ecommerce requirement)

File: app/config/sonata/sonata_block.yml (hunk 3 above)
Comment (wbloszyk) [reply app/config/sonata/sonata_block.yml]: Not only this block are auto-registred. I just delete duplicated config.

File: app/config/config.yml [hunk 5]
```diff
@@ -127,14 +128,14 @@ jms_serializer:
 
 # Assetic Configuration
 #assetic:
```
Comment (core23) [inline app/config/config.yml]: You can remove the `assetic` part. It is deprecated by symfony

File: app/config/services.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        # if you need to do this, you can override this setting on individual services
+        public: false
+
+    # makes classes in src/AppBundle available to be used as services
+    # this creates a service per class whose id is the fully-qualified class name
+    AppBundle\:
+        resource: '../../src/AppBundle/*'
+        # you can exclude directories or files
+        # but if a service is unused, it's removed anyway
+        exclude: '../../src/AppBundle/{Entity,Repository,Tests}'
```
Comment (core23) [inline app/config/services.yml]: Do we have Test classes under the `src` directory?

File: app/config/sonata/sonata_block.yml [hunk 7]
```diff
@@ -55,13 +55,9 @@ sonata_block:
...
             cache: sonata.page.cache.js_async
             contexts: [user]
-        sonata.basket.block.nb_items: # FixMe: Why is this duplicated?
```
Comment (core23) [inline app/config/sonata/sonata_block.yml]: You can remove every listed block here, as long as it does not provide any non-standard configuration. Block services are autoloaded since the last minor version.

File: bin/qa_behat.sh [hunk 8]
```diff
@@ -10,14 +10,15 @@ if [ ! -d "build" ]; then
     mkdir build
 fi
 
-parameters=" -f progress,junit --out ,build/junit"
+#parameters=" -f pretty --out build/junit"
+parameters=" "
 
 echo "Running Behat default test suite"
-bin/behat -f progress $parameters
+vendor/bin/behat -f progress $parameters
```
Comment (core23) [inline bin/qa_behat.sh]: Why this change? AFAIK the file does still exist under `bin`

File: composer.json [hunk 9]
```diff
@@ -5,12 +5,26 @@
...
+            "app/AppKernel.php",
+            "app/AppCache.php"
+        ]
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "Tests\\": "tests/",
+            "Sonata\\Bundle\\QABundle\\Tests\\": "src/Sonata/Bundle/QABudle/Tests/"
+        },
+        "files": [
```
Comment (core23) [inline composer.json]: Do we need this?

File: composer.json [hunk 10]
```diff
@@ -5,12 +5,26 @@
...
         "classmap": [
-            "app/AppKernel.php"
+            "app/AppKernel.php",
+            "app/AppCache.php"
+        ]
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "Tests\\": "tests/",
+            "Sonata\\Bundle\\QABundle\\Tests\\": "src/Sonata/Bundle/QABudle/Tests/"
```
Comment (core23) [inline composer.json]: Can you move the tests to the new `/tests` directory

File: bin/qa_behat.sh (hunk 8 above)
Comment (wbloszyk) [reply bin/qa_behat.sh]: qa_behat.sh is script for running behat with some config.
Pure vendors script like behat, phpunit ... was change from bin/* to vendor/bin/*

File: app/config/config.yml (hunk 5 above)
Comment (wbloszyk) [reply app/config/config.yml]: I will do it in next PR. Also then i will add symfony/webpack-encore.

File: app/config/services.yml (hunk 6 above)
Comment (wbloszyk) [reply app/config/services.yml]: No. I make this config for fast create working sandbox (at first ecommerce 3.0 must be stable). All tests will be in tests/ directory.

File: app/config/sonata/sonata_block.yml (hunk 7 above)
Comment (wbloszyk) [reply app/config/sonata/sonata_block.yml]: Thx for this information.

File: composer.json (hunk 9 above)
Comment (wbloszyk) [reply composer.json]: For now it needed. I just run "vendor/bin/phpunit src/". It's good idea to remove QABundle test to tests/, then this config can be removed.

File: composer.json (hunk 10 above)
Comment (wbloszyk) [reply composer.json]: Ok, I just write it one comment ago :)

Comment (wbloszyk) [reply composer.json]: Anyway i will do it after make ecommerce stable. Then all test can be checked.

**Comment by wbloszyk** [comment]
@core23 @greg0ire @salamichel  
Can u review it? I need some feedback.

1. What do you think about block sonata.user.account.block? Should be add to DemoBundle or SonataUserBundle like feature?
1. Can you update sandbox and dev-kit? 
So many files need ```declare(strict_types=1);```. Configure it like ecommerce will be nice.

@salamichel
You can confirm your [#644](https://github.com/sonata-project/ecommerce/issues/633) issue, becouse sandbox have also this bug.

File: app/config/config.yml [hunk 11]
```diff
@@ -147,4 +147,4 @@ nelmio_api_doc: ~
 sensio_framework_extra:
     view:    { annotations: true } # FOS\RestBundle\DependencyInjection\Compiler\ConfigurationCheckPass: You must enable the SensioFrameworkExtraBundle view annotations to use the ViewResponseListener
     router:  { annotations: true }
-    request: { converters:  true }
+    request: { converters:  true }
```
Comment (greg0ire) [inline app/config/config.yml]: Please configure your IDE to always add a newline at end of file, this will make the red pictogram disappear.

File: app/config/config_dev.yml [hunk 12]
```diff
@@ -42,8 +42,8 @@ monolog:
         #    type:   chromephp
         #    level:  info
 
-assetic:
-    use_controller: true
+#assetic:
+#    use_controller: true
```
Comment (greg0ire) [inline app/config/config_dev.yml]: To remove entirely?

File: behat.yml.dist [hunk 13]
```diff
@@ -22,4 +22,63 @@ api:
...
+    suites:
+        paths:
+            features:  features/api
+            bootstrap: features/bootstrap
+        filters:
+            tags: "~@skipped&&@api"
+wip:
+    suites:
+        filters:
+            tags: "@skipped"
```
Comment (greg0ire) [inline behat.yml.dist]: missing newline

File: app/config/config.yml (hunk 11 above)
Comment (wbloszyk) [reply app/config/config.yml]: I used php-cs-fixer, it should be fixed. (the same issue i have in xml). Thanks for advise.

File: behat.yml.dist (hunk 13 above)
Comment (wbloszyk) [reply behat.yml.dist]: Fix too.

File: app/config/config_dev.yml (hunk 12 above)
Comment (wbloszyk) [reply app/config/config_dev.yml]: I will delete all assetic deprecation when I change assetic to assets or symfony/webpack-encore

**Comment by wbloszyk** [comment]
@greg0ire @soullivaneuh @core23 I'm done. Feel free to review it.


It is more like alfa version becouse some things must be fix:
- phpunit tests (after update phpunit)
- product comment do not work (issue in SonataCommentBundle)
- product add modal do not appare,
- get "Operation timed out" in guzzle after sendbankAction

**Comment by core23** [comment]
There are a lot of changes in this big commit. Reviewing this whole bunch is hard. Can you try to split this into some logical parts @wbloszyk ?

**Comment by wbloszyk** [comment]
I will try.

**Comment by marcinmarzeta** [comment]
@wbloszyk  
Hello Wojtek,
Do you need any help with this?

**Comment by wbloszyk** [comment]
@marcinmarzeta I do not need help. Now I have holiday. I will back on monday. I working also on sandbox:4.0, where you can help with webpack or API issue.

@core23 @greg0ire Can you add new 4.0 branch?

**Comment by greg0ire** [comment]
Sure, done!

**Comment by wbloszyk** [comment]
@core23 I split it. Mybe not perfect, but now is readable.

File: .php_cs.dist [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+ /**
+ * DO NOT EDIT THIS FILE!
+ *
+ * It's auto-generated by sonata-project/dev-kit package.
+ */
+
+ $header = <<<'HEADER'
+This file is part of the <name> project.
+
+(c) <yourname> <youremail>
```
Comment (core23) [inline .php_cs.dist]: This is right?

File: tests/Sonata/Bundle/QABundle/CommandTestCase.php [hunk 15]
```diff
@@ -11,7 +11,7 @@
  * file that was distributed with this source code.
  */
 
-namespace Sonata\Bundle\QABundle\Tests;
+namespace Tests\Sonata\Bundle\QABundle;
```
Comment (core23) [inline tests/Sonata/Bundle/QABundle/CommandTestCase.php]: Why do we use `Tests` as root prefix? We don't use this in any other project.

File: .php_cs.dist (hunk 14 above)
Comment (wbloszyk) [reply .php_cs.dist]: I don't know. I check 2.3 branch where DemoBundle and QABundle have `Sonata package`. Other part had `<name> project`. I think for sandbox we should use `<name> project`. 

Should I and new rules and finder to generate two header version, one for project and one for Demo + QA Bundles?

File: tests/Sonata/Bundle/QABundle/CommandTestCase.php (hunk 15 above)
Comment (wbloszyk) [reply tests/Sonata/Bundle/QABundle/CommandTestCase.php]: Sandbox is symfony(here 3.4) project, not bundle, and have this structure. Also all tests from this bundle are for project, not for bundle test. 

After upgrade Symfony to 4.x and flex DemoBundle(and QABundle) should be merge to App or move to own repository.

**Comment by core23** [comment]
> @core23 I split it. Mybe not perfect, but now is readable.

To make the review and merge process even faster, can you create some atomic PRs (e.g. fix cs) so we can merge some parts soon and this PR is not blocked until all changes are applied.

**Comment by core23** [comment]
Here are some ideas for the next PRs:
- Refactor tests (PHPUnit, behat)
- Refactor project
- Add UserBundle
- Update documentation

**Comment by wbloszyk** [comment]
@Contributors 
Another step is ready to check: #623 

Next will be: "Refactor tests"
